packages/@aws-cdk/aws-rds/lib/props.ts

```diff
@@ -97,27 +97,142 @@ export interface BackupProps {
 /**
  * Username and password combination
  */
-export interface Login {
+export abstract class Login {
```

Comment:
I also prefer a concrete class here... with the static methods returning a `new Login()`... there's something similar here

https://github.com/aws/aws-cdk/blob/94bbabfff8b785656a40e85539f97ae8cb83c818/packages/%40aws-cdk/aws-amplify/lib/basic-auth.ts#L50-L53

I also wonder if `MasterCredentials` or just `Credentials` is a better name...

Reply:
> I also prefer a concrete class here...

Out of curiosity, why? Just stylistic preference, or is there something else I'm missing?

I'm not strongly opposed; however, it's worth noting that changing `Login` from an interface (today) to a concrete class _will_ be a breaking change for all active Cluster users, meaning this PR will break *all* current RDS users (both Instance and Cluster setups). As an abstract class, that's not the case. If we're going to break, it does need to be now, but I'm not sure the minor aesthetics of the class being concrete are worth it.

Reply:
If you go from this
```ts
interface Login {
  readonly username: string;
  readonly password?: string;
  readonly encryptionKey?: kms.IKey;
}
```

to this:
```ts
class Login {
  public static fromUsername(username: string, options: FromUsernameOptions = {}): Login {
    return new Login(username, options.password, options.encryptionKey);
  }

  public static fromSecret(secret: secretsmanager.Secret): Login {
    return new Login(
      secret.secretValueFromJson('username').toString(),
      secret.secretValueFromJson('username'),
      secret.encryptionKey,
      secret,
    );
  }

  public readonly username: string;
  public readonly password?: SecretValue;
  public readonly encryptionKey?: kms.IKey;

  constructor(username: string, password?: SecretValue, encryptionKey?: kms.IKey, secret?: secretsmanager.Secret) {
    this.username = username;
    this.password = password;
    this.encryptionKey = encryptionKey;
    this.secret = secret;
  }
}
```

it is not breaking, at least in TS, no?

Reply:
Ah, you're right. I had done a slightly different implementation (getters) that didn't work; this approach is fine (at least in Typescript).

I'm still curious if the perceived advantage of this approach is stylistic, or if there's a benefit I'm missing.

Reply:
+1 for `credentials` or `masterCredentials` as the name of the property, that's a great suggestion @jogold !